### PR TITLE
Add WebTransport datagram support.

### PR DESCRIPTION
**Pull request by jianjunz**
This change adds RTP packtiezer to QUIC agent. QUIC agent is able to
send RTP packets over WebTransport datagrams.

### CODE REVIEW

**Comment by jianjunz** [comment]
Converted to draft since client changes are not ready.

**Comment by jianjunz** [comment]
Hi, @starwarfan, @qwu16, this change is ready for review. Please take a look, thanks.

File: source/agent/addons/quic/QuicTransportConnection.cc [hunk 1]
```diff
@@ -137,10 +161,38 @@ NAN_METHOD(QuicTransportConnection::close)
...
+{
+    m_asyncOnClose.data = this;
+    uv_async_send(&m_asyncOnClose);
+}
+
+void QuicTransportConnection::OnDatagramReceived(const uint8_t* data, size_t length)
+{
+    owt_base::FeedbackMsg feedback = { .type = owt_base::DATA_FEEDBACK, .cmd = owt_base::RTCP_PACKET };
+    feedback.buffer.len = length;
+    memcpy(feedback.buffer.data, data, length);
```
Comment (starwarfan) [inline source/agent/addons/quic/QuicTransportConnection.cc:196]: better check length here

File: source/agent/addons/quic/QuicTransportConnection.cc [hunk 2]
```diff
@@ -137,10 +161,42 @@ NAN_METHOD(QuicTransportConnection::close)
...
+}
+
+void QuicTransportConnection::onFrame(const owt_base::Frame& frame)
+{
+    if (frame.format != owt_base::FRAME_FORMAT_RTP) {
+        ELOG_WARN("WebTransport datagram output only supports RTP packets. Received frame type %d. Discard this frame.", frame.format);
+        return;
+    }
+    if (!m_session) {
+        ELOG_WARN("WebTransport session is nullptr.");
```
Comment (starwarfan) [inline source/agent/addons/quic/QuicTransportConnection.cc:179]: It's unexpected, but better not put `m_session->SendOrQueueDatagram(frame.payload, frame.length);` after `!m_session`

File: source/agent/addons/quic/WebTransportFrameDestination.cc [hunk 3]
```diff
@@ -45,4 +54,83 @@ NAN_METHOD(WebTransportFrameDestination::newInstance)
     WebTransportFrameDestination* obj = new WebTransportFrameDestination();
     obj->Wrap(info.This());
     info.GetReturnValue().Set(info.This());
+}
+
+NAN_METHOD(WebTransportFrameDestination::addDatagramOutput)
+{
+    WebTransportFrameDestination* obj = Nan::ObjectWrap::Unwrap<WebTransportFrameDestination>(info.Holder());
+    if (obj->m_datagramOutput) {
```
Comment (starwarfan) [inline source/agent/addons/quic/WebTransportFrameDestination.cc:62]: Should you call `obj->m_datagramOutput->FrameDestination()->setDataSource(nullptr);` for previous datagramOutput here?

File: source/agent/addons/quic/QuicTransportConnection.cc (hunk 2 above)
Comment (jianjunz) [reply source/agent/addons/quic/QuicTransportConnection.cc:179]: Added return if `m_session` doesn't exist.

File: source/agent/addons/quic/WebTransportFrameDestination.cc (hunk 3 above)
Comment (jianjunz) [reply source/agent/addons/quic/WebTransportFrameDestination.cc:63]: Done.

File: source/agent/addons/quic/WebTransportFrameDestination.cc [hunk 4]
```diff
@@ -45,4 +54,83 @@ NAN_METHOD(WebTransportFrameDestination::newInstance)
...
+    NanFrameNode* output = Nan::ObjectWrap::Unwrap<NanFrameNode>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
+    obj->m_datagramOutput = output;
+    // TODO: Use addDataDestination defined in MediaFramePipeline. We use m_datagramOutput here because the output could also be streams.
+    output->FrameDestination()->setDataSource(obj);
+}
+
+NAN_METHOD(WebTransportFrameDestination::removeDatagramOutput)
+{
+    WebTransportFrameDestination* obj = Nan::ObjectWrap::Unwrap<WebTransportFrameDestination>(info.Holder());
+    std::unique_lock<std::shared_timed_mutex> lock(obj->m_datagramOutputMutex);
```
[the remark this answers is not in the document]
Comment (jianjunz) [reply source/agent/addons/quic/WebTransportFrameDestination.cc:80]: Yes, added lock to avoid races.

File: source/agent/quic/index.js [hunk 5]
```diff
@@ -161,19 +172,29 @@ module.exports = function (rpcClient, selfRpcId, parentRpcId, clusterWorkerIP) {
...
-    const createStreamPipeline =
-        function(streamId, direction, options, callback) {
+    const createFrameDestination = (subscriptionId) => {
+      if (frameDestinationMap.has(subscriptionId)) {
+        log.warn('Frame destination for ' + subscriptionId + ' exists.');
+        return;
+      }
+      const frameDestination =
+          new addon.WebTransportFrameDestination(subscriptionId);
+      frameDestinationMap.set(subscriptionId, frameDestination);
```
Comment (starwarfan) [inline source/agent/quic/index.js:189]: You could clean `frameDestinationMap` in the future.

**Review by starwarfan — APPROVED** [review]
Better add clean up code in later commit

**Comment by jianjunz** [comment]
Sure. #1083 was filed for clean up.

**Pull request merged**